Set default for DeploymentConfigSpec.RevisionHistoryLimit in apps/v1

tnozicka opened the pull request:
At this point we don't default DeploymentConfigSpec.RevisionHistoryLimit in apps/v1 which is causing the deployments to have unlimited history which will lead to performance issues at scale and when deploying regularly. 

Note that upstream also defaulted RevisionHistoryLimit to 10 for Deployments in https://github.com/kubernetes/kubernetes/pull/49924

Needs to be backported to 3.6.1 as well.

cc: @smarterclayton @mfojtik @kargakis

tnozicka commented:
@openshift/api-review

kargakis commented:
/approve
/lgtm

tnozicka commented:
infra failed 
/test unit

kargakis commented:
/retest

tnozicka commented:
/retest

tnozicka commented:
/retest

smarterclayton commented:
My email never got delivered it looks like.  The defaults can only be applied when creating via the "apps.openshift.io" group, not the default group.  There needs to be specific code to check for that just like: https://github.com/openshift/origin/blob/master/pkg/cmd/server/origin/legacy.go#L190

tnozicka commented:
@smarterclayton my bad :( It was already like 9 PM on Friday for me and I was misguided by defaulters located under path `deploy/apis/apps/v1/defaults.go`. Should have tested creating resources in both API groups though...

I am feeling a bit hesitant to do this at storage layer but I might be wrong. We have defaulters, right? But at this point we register the same ones for legacy and the new group https://github.com/openshift/origin/blob/7580da21a2133f7d3e013b1bfe137db33110f4d3/pkg/deploy/apis/apps/v1/register.go#L19-L22

Shouldn't I better plumb it there?

smarterclayton commented:
The line I pasted is where the storage is customized one way or another.
The strategy is what encapsulates the additional defaulting, so it's
probably better that we keep the registration of the defaulting there.
Creating net new types has a bunch of other impacts that we're probably not
ready for (but which at some point will become necessary).  When we do
duplicate the types, we'll change how we register.  Too big for the near
term though.

On Tue, Aug 8, 2017 at 6:16 AM, Tomáš Nožička <notifications@github.com>
wrote:

> @smarterclayton <https://github.com/smarterclayton> my bad :( It was
> already like 9 PM on Friday for me and I was misguided by defaulters
> located under path deploy/apis/apps/v1/defaults.go. Should have tested
> creating resources in both API groups though...
>
> I am feeling a bit hesitant to do this at storage layer but I might be
> wrong. We have defaulters, right? But at this point we register the same
> ones for legacy and the new group https://github.com/openshift/
> origin/blob/7580da21a2133f7d3e013b1bfe137db33110f4d3/pkg/deploy/apis/
> apps/v1/register.go#L19-L22
>
> Shouldn't I better plumb it there?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/15642#issuecomment-320914144>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p1ryEttb5vKHq_lqyXgIfqj8pxVPks5sWDWJgaJpZM4OuDCi>
> .
>

tnozicka commented on pkg/cmd/server/origin/legacy.go:
```diff
@@ -206,7 +207,8 @@ func LegacyStorage(storage map[schema.GroupVersion]map[string]rest.Storage) map[
 				case "deploymentConfigs":
 					restStorage := s.(*deploymentconfigetcd.REST)
 					store := *restStorage.Store
-					restStorage.DeleteStrategy = orphanByDefault(store.DeleteStrategy)
```
I have my doubts about if this ever worked - `restStorage` should have been `store`

tnozicka replied:
@smarterclayton @mfojtik This seems to be indeed a bug that we have carried for a long time.

What do we want to do with legacy API cascade deletion? We enabled it for new DCs by https://github.com/openshift/origin/blob/e98a1a9bcf47784533bf76f2d1cae7659b5b4acd/pkg/deploy/controller/deploymentconfig/controller_ref_manager.go#L143 but fixing this bug seems to mitigate it. You can see it broken in our test using the legacy client: https://github.com/openshift/origin/blob/e98a1a9bcf47784533bf76f2d1cae7659b5b4acd/test/extended/deployments/deployments.go#L1098 https://ci.openshift.redhat.com/jenkins/job/test_pull_request_origin_extended_conformance_install_update/4085/testReport/junit/(root)/Extended/deploymentconfigs__should_adhere_to_Three_Laws_of_Controllers__Conformance_/

tnozicka replied:
I was wrong that the patch would change the propagation policy. (We used to set a finalizer here, but now there is only blockOwnerDeletion which doesn't affect it in that way. So I will rephrase the question:

@smarterclayton @mfojtik I guess we initially wanted DCs created by legacy API not to have cascading deletion but given that we have failed at it due to this bug. Given that it has already been in a release (https://github.com/openshift/origin/blame/master/pkg/cmd/server/origin/legacy.go#L209) do we want to turn it back again? What's the policy for that?

tnozicka replied:
@mfojtik @smarterclayton bump

mfojtik replied:
@tnozicka that is right, it should be `store` and not `restStorage` (@deads2k).

I think we should fix this bug and backport to 3.5, @deads2k @smarterclayton thoughts?

tnozicka commented:
I wanted to wrap the defaulting functions just for [group API SchemeBuilder](https://github.com/openshift/origin/blob/7580da21a2133f7d3e013b1bfe137db33110f4d3/pkg/deploy/apis/apps/v1/register.go#L22) but that didn't work because they get called even for legacy resources (it calls defaulting functions twice for legacy resources) so I ended up modifying it on storage layer.

I am not excited to have defaulting scattered between 2 places but this is probably the only option here. And BuildConfigs do the same.

I will add an integration test tomorrow but this is mostly done.

smarterclayton commented on pkg/deploy/registry/deployconfig/strategy.go:
```diff
@@ -105,12 +114,44 @@ func (strategy) CheckGracefulDelete(obj runtime.Object, options *metav1.DeleteOp
+
+	dc := obj.(*deployapi.DeploymentConfig)
+
+	if dc.Spec.RevisionHistoryLimit == nil {
```
Split this into its own method for sanity.  `func AppsV1DeploymentConfigLayeredDefaults(*deployapi.DeploymentConfig)` and add a good godoc comment.

smarterclayton commented:
Looks good to me, go ahead and squash.

tnozicka commented:
yum repo flake
/retest

smarterclayton commented:
/retest

tnozicka commented:
I will need to plumb in the new group client for tests and use that in 
https://github.com/openshift/origin/blob/e98a1a9bcf47784533bf76f2d1cae7659b5b4acd/test/extended/deployments/deployments.go#L1093-L1098 
instead of the legacy client - that's why the `extended_conformance_gce` test is failing

tnozicka commented:
- rebased to master
- plumbed in the group API into testing framework (only for AppsV1 but in a general and extensible way
- fixed the cascade delete test using the group API

smarterclayton commented:
Looks perfect

/lgtm

tnozicka commented:
Seems like some changes in code generation went into master from last time I run `make update`.
Updated files:
```
	modified:   pkg/user/generated/informers/externalversions/generic.go
	new file:   pkg/user/generated/informers/externalversions/user/v1/identity.go
	modified:   pkg/user/generated/informers/externalversions/user/v1/interface.go
	modified:   pkg/user/generated/informers/internalversion/generic.go
	new file:   pkg/user/generated/informers/internalversion/user/internalversion/identity.go
	modified:   pkg/user/generated/informers/internalversion/user/internalversion/interface.go
	modified:   pkg/user/generated/listers/user/internalversion/expansion_generated.go
	new file:   pkg/user/generated/listers/user/internalversion/identity.go
	modified:   pkg/user/generated/listers/user/internalversion/user.go
	modified:   pkg/user/generated/listers/user/v1/expansion_generated.go
	new file:   pkg/user/generated/listers/user/v1/identity.go
```
no other changes

tnozicka commented:
/retest

tnozicka commented:
Another change in code generation in master - rebased and added new generated files:
```
	modified:   api/docs/apis-apps.openshift.io/v1.DeploymentConfig.adoc
	modified:   api/docs/oapi/v1.DeploymentConfig.adoc
```

mfojtik commented:
/lgtm

tnozicka commented:
@kargakis not that I want to complain but submit queue probably shouldn't merge PRs with **needs-api-review** label
